### ISSUE TITLE
Security aspect

### ISSUE THREAD

**Issue by Auronius** [karagog/Gryptonite#77]
Good Day!
I have been using your password manager Gryptonite for two days only, but I already amazed, how could you make so simple, fast but powerful application, which does exactly what it should do, without all this unnecessary stuff (for me) like cloud storage, WiFi synchronization etc... and without heavy graphical interface!
**Excellent work!**
Since KeePassX developers have made KeePassX v.2, I started to search another password manager, because KeePassX 2 is how to say polite - a very bad joke for those who knew KeePassX v.1. And looks like I found just what I needed!
Nevertheless, I have some questions about Gryptonite's security, because I am planning to use it on daily basis for critical data like important e-mail addresses, bank account, PayPal, web-shops, so security aspect is **very** important for me. By the way, that is the main reason I have switched to Linux OS. Personally I know very little about Cryptography, mostly from wiki, still I am trying to compare Gryptonite's security with KeePass and LastPass security.
1. As I can understand Gryptonite uses AES Rijndael encryption algorithm with block size of 128 bits, but which key length it uses? 256 bits with 14 cycles of repetition?
2. Which cryptographic hash function is used for compressing the components of master key? SHA-256?
3. Have you implemented a protection against dictionary and guessing attacks, like PBKDF2 in LastPass or AES-KDF in KeePass? If yes, how many iterations Gryptonite performs to produce a derived key?
4. Does sensitive data behind asterisks is stored encryptedly in process memory?
5. Does Gryptonite performs quick self-test on start to see whether the encryption and hash algorithms work correctly and pass their test vectors?
6. What is CCM?
7. How to change master password?
8. Do you still maintain Gryptonite?
**Again, thank You for your work!**

**Comment by karagog** [comment]
Auronius,

Thanks for the glowing feedback, I'm glad you like my application :) You hit on several pain points I experienced when adopting a password manager years ago. The only real contender then was KeePass, and I was immediately turned off by the interface - it was like it was designed for hundreds of different things, with very little attention paid to the majority user - the one who just wants to store their secrets simply and securely. I was primarily an embedded engineer back then, so I decided to use this project to learn about UI development and application architecture, so I got great practice and made an incredibly useful thing that I can share with the world - it's a win-win!

While making Gryptonite, I learned a lot myself about encryption and storing secrets safely. The initial version was very naive and had several major security flaws. Subsequently, I read more about encryption on the internet and realized all of the flaws. I then redesigned/rearchitected the entire application from the ground up into its current state today. It has not been subjected to a thorough security review, but I believe it to be highly secure. I am probably as paranoid as you are about security, so hopefully my responses to the questions below are satisfactory:

1. As I can understand Gryptonite uses AES Rijndael encryption algorithm with block size of 128 bits, but which key length it uses? 256 bits with 14 cycles of repetition?

The key size is 256 bits (which implies 14 cycles of repetition as an implementation detail of AES).

2. Which cryptographic hash function is used for compressing the components of master key? SHA-256?

It uses the more modern SHA3-256. Details of the key-derivation function are in the code at: https://github.com/karagog/gutil/blob/c4e319ea43b213266e6ca3ef04097504e46b2f79/src/cryptopp/cryptopp_cryptor.cpp#L320

The data encryption function is at: https://github.com/karagog/gutil/blob/c4e319ea43b213266e6ca3ef04097504e46b2f79/src/cryptopp/cryptopp_cryptor.cpp#L132

3. Have you implemented a protection against dictionary and guessing attacks, like PBKDF2 in LastPass or AES-KDF in KeePass? If yes, how many iterations Gryptonite performs to produce a derived key?

Yes, the key-derivation function does 65536 iterations. For extra protection you can consider using a keyfile + master password, and storing the keyfile on a USB drive or something (maybe keep a backup somewhere safe too, b/c if you lose it you lose everything!)
https://github.com/karagog/gutil/blob/c4e319ea43b213266e6ca3ef04097504e46b2f79/src/cryptopp/cryptopp_cryptor.cpp#L31

4. Does sensitive data behind asterisks is stored encryptedly in process memory?

No, the asterisks are purely a visual distinction. They only provide protection against someone looking over your shoulder, not peering into process memory. I would need convincing that it's worthwhile, because I tend to think there's no good way to even do that, given that you interact with the secrets in plaintext (e.g. when you copy to clipboard) so it's going to be exposed as plaintext in-memory eventually. Let's just say you should be able to trust your own machine when using Gryptonite, but that's true of almost any application I think.

5. Does Gryptonite performs quick self-test on start to see whether the encryption and hash algorithms work correctly and pass their test vectors?

No. I imagine this could be useful for diagnostics, but if the crypto layer broke I think it would be immediately obvious to the user, and would not lead to corrupted data or things of that nature. The risk is if a low-level library is installed on the machine that breaks encryption, but this will prevent you from opening the file at all (because the key-check will fail) so there is no risk of corrupting file data from a broken encryption layer.

Admittedly, this is probably necessary for a full-fledged professional solution, but I'm just one man working for free in his living room - so it's probably never going to happen.

6. What is CCM?

https://en.wikipedia.org/wiki/CCM_mode

It simply describes the block-chaining mode for encryption. AES is the cipher used to encrypt a single block of data (e.g. 256 bytes) but CCM provides the algorithm for encrypting several blocks of data. Or, for example, encrypting odd sizes like 257 bytes. One major feature of CCM is that it provides authentication. A message is authentic if it has not been tampered with by anyone - either maliciously or accidentally (e.g. bit-rot). A bad actor can flip a bit on your secrets file which corrupts the database, but they cannot flip a bit without us knowing about it. With other modes you might successfully decrypt the message but it looks like garbage - with CCM we get an authoritative decryption error. BTW I could have chosen any one of several block-chaining modes (see GCM), but this was a design choice I made after a lot of careful thought.

7. How to change master password?

File -> Save As
(then delete the old database)

8. Do you still maintain Gryptonite?
Not actively. I wrote it many years ago when I was young and bored, but now that my talents have been discovered, I'm at a big tech company and no longer have time for open-source development. I would like to hand it off to the next generation of developers and serve as a consultant/reviewer, but so far nobody has come forward to take it over. Do you know anyone? ;)

**Comment by Auronius** [comment]
All right. Now I can see, that Gryptonite has strong security and can be used for storing sensitive data, just what I need. Unfortunately, I don't know any programmer, who could continue maintaining Your project. I haven't tested Gryptonite extensively, but for now, I think it has all necessary functions to operate as password manager, so no extra stuff needed. As for me, it should be left as it is, except one important thing.
You have written, that the key-derivation function performs 65536 iterations. I think, nowadays it is very small number, because, as you know, PC and CPU power increases geometrically and programmers create new methods of performing dictionary and guessing attacks, such as GPU/ASIC attacks. So if it is not difficult, could You please increase number of iterations till **900000** or **one million**? As I understand this will greatly increase database security, by making much harder to perform brute-force attacks. For instance, I still have KeePassX installed on my PC with Intel Pentium G2030, and 4GB RAM, and it can perform 8300000 iterations per 1 second, so I don't think, that increasing iteration count to one million will noticeably reduce Gryptonite's performance on majority of PC's.
![transformation_rounds](https://cloud.githubusercontent.com/assets/13540006/22063204/3b96f046-dd86-11e6-9be5-873abce13ecf.png)

**Comment by karagog** [comment]
I came up with the value of 65536 a few years ago after calibrating it manually. I started small and kept doubling it until the amount of time required to generate a key would be an unacceptable delay to a typical end-user (this unacceptable value was 2x65536, which is why I settled on 65536).

The "number of iterations" I believe can be misleading, because it depends on the algorithm used to generate the hash. I don't know what "transform rounds" mean (from your screenshot), but it's probably some implementation-specific detail for KeePass. In the case of Gryptonite, the # of iterations means that the hash is generated by re-hashing the previous hash ~65K times. It is perhaps possible that someone will come up with a faster way of doing a SHA3 hash and then this limit will be insecure. It is also possible that CPU/GPU power will increase to the point where regular off-the-shelf hardware would be able to brute-force it eventually. Also consider the fact that not everyone has high-end hardware, and that by increasing the # of iterations it could render Gryptonite unworkable for large numbers of people.

At the time of development, I decided the current risk is acceptable, because of the fact that Gryptonite is supposed to be **very user friendly**, and that by adding more bells and whistles to the KDF it increases complexity for the end-user only to satisfy the whims of a very small segment of users (this is what happened to KeePass, IMHO).

Another thing to consider is that by changing the # of iterations, that would require a non-backwards compatible update to the file format, because all the old files would be unreadable to the new software version. It is totally possible to do, but it's not easy and requires a lot of development effort to make sure this transition happens smoothly, without incurring any data loss for users (something that keeps me up at night). I already suffered through this when updating file formats several times in the past (between every big version number update), so I know intimately what is required, and unfortunately I don't have the cycles to spend on this anymore, except in a consulting role.

If someone needs protection from government agencies trying to use supercomputers to brute-force dictionary attack them, then perhaps Gryptonite is not for them. I wrote it targeting a different segment of users - those who want a simple and delightful user experience while still maintaining a very high level of security. We can re-evaluate this KDF iterations value over time, but IMO it should be targeted to satisfy the ~90th percentile of users.
